roles/ceph-osd/tasks/check_mandatory_vars.yml

```diff
@@ -18,114 +24,88 @@
-    - not osd_scenario == "lvm"
     - devices|length == 0
 
 - name: check if osd_scenario lvm is supported by the selected ceph version
   fail:
     msg: "osd_scenario lvm is not supported by the selected Ceph version, use Luminous or newer."
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
     - not containerized_deployment
-    - osd_scenario == "lvm"
+    - lvm
     - ceph_release_num.{{ ceph_release }} < ceph_release_num.luminous
 
 - name: verify osd_objectstore is 'filestore' when using the lvm osd_scenario
   fail:
     msg: "the lvm osd_scenario currently only works for filestore, not bluestore"
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
-    - osd_scenario == "lvm"
+    - lvm
     - not osd_auto_discovery
     - osd_objectstore != 'filestore'
 
 - name: verify lvm_volumes have been provided
   fail:
     msg: "please provide lvm_volumes to your osd scenario"
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
-    - osd_scenario == "lvm"
+    - lvm
     - not osd_auto_discovery
     - lvm_volumes|length == 0
 
 - name: make sure the lvm_volumes variable is a list
   fail:
     msg: "lvm_volumes: must be a list"
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
+    - lvm
     - not osd_auto_discovery
-    - osd_scenario == "lvm"
     - lvm_volumes is string
 
 - name: make sure the devices variable is a list
   fail:
     msg: "devices: must be a list, not a string, i.e. [ \"/dev/sda\" ]"
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
-    - not osd_auto_discovery
+    - collocated or non_collocated
     - devices is string
 
-- name: verify journal devices have been provided
+- name: make sure the dedicated_devices variable is a list
   fail:
-    msg: "please provide devices and raw journal devices to your osd scenario"
+    msg: "dedicated_devices: must be a list, not a string, i.e. [ \"/dev/sda\" ]"
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
     - not containerized_deployment
-    - osd_scenario == 'non-collocated'
-    - dedicated_devices|length == 0
-      or devices|length == 0
+    - non_collocated
+    - dedicated_devices is string
 
-- name: make sure the dedicated_devices variable is a list
+- name: verify journal devices have been populated in non-auto discovery
   fail:
-    msg: "dedicated_devices: must be a list, not a string, i.e. [ \"/dev/sda\" ]"
+    msg: "specify discovery option or pre-populate ceph journal devices in non-collocated osd scenario"
   when:
-    - osd_group_name is defined
-    - osd_group_name in group_names
     - not containerized_deployment
-    - osd_scenario == 'non-collocated'
-    - dedicated_devices is string
+    - non_collocated
+    - not osd_auto_discovery
     - dedicated_devices|length == 0
-      or devices|length == 0
+
+- name: verify auto discovery populates devices and journal devices
```

Comment:
The task which populates `devices` and `dedicated_devices` when `osd_auto_discovery` is `True` is played after this check.

Reply:
Yes.  Setting osd_auto_discovery to True causes the discover_devices.yml to probe the ansible_devices and populates the devices and then dedicated_devices.  Setting the osd_auto_discovery to False requires manual entry for the devices and dedicated_devices lists.

Reply:
@AlLau yes this is what I understood, so if I set `osd_auto_discovery` to `True` I expect `devices` and `dedicated_devices` to be automatically populated as you said.
My question is how can I pass the check I mentioned above with `osd_auto_discovery: True` since that check which is in charge to stop the playbook if `devices` and/or `dedicated_devices` are not defined/populated is played before `discover_devices.yml`?

Reply:
@guits In the defaults/main.yml, the osd_auto_discovery, devices, dedicated_devices are defined with defaults.  In the check on line 97 in check_mandatory_vars.yml, if the osd_auto_discovery is set to True, it expects devices and dedicated_devices to be empty lists ([]) to start with depending on the chosen osd_scenario.  The devices and dedicated_devices are then populated in discover_devices.yml.  In tasks/main.yml, check_mandatory_vars.yml is called before discover_devices.yml, so the check in line 97 passes.  The idea is that when osd_auto_discovery is set to True, the devices and dedicated_devices are empty lists to begin with.

Reply:
Currently, osd_auto_discovery is defaulted to false in defaults/main.yml.  The devices and dedicated_devices are defaulted to [].  In our production use, the osd_auto_discovery is defaulted to true.

The osd_auto_discovery did not evaluated to True in line 101 in check_mandatory_vars.yml.  So, the fail did not fired.

I would like to suggested stick with the meaning of osd_auto_discovery:

if osd_auto_discovery is true, devices and dedicated_devices are empty.  Make this the default.

If osd_auto_discovery is false, devices and dedicated_devices are populated somewhere in the inventory before check_mandatory_vars.yml gets executed.

In our production environment, we saw cases where the osd_auto_discvoery is set to true and yet the devices and dedicated_devices are populated.  Because of this, those checks were put in place.

Reply:
If this did not answer your question, I will try again.